packages/vm/lib/buildBlock.ts

```diff
@@ -0,0 +1,250 @@
+    for (const [i, txResult] of this.transactionResults.entries()) {
+      const tx = this.transactions[i]
+      gasUsed.iadd(txResult.gasUsed)
+      const { encodedReceipt } = await generateTxReceipt.bind(this.vm)(tx, txResult, gasUsed)
+      await receiptTrie.put(encode(i), encodedReceipt)
+    }
+    return receiptTrie.root
+  }
+
+  /**
+   * Rewards the miner.
+   */
+  private async rewardMiner() {
+    const minerReward = new BN(this.vm._common.param('pow', 'minerReward'))
+    const reward = calculateMinerReward(minerReward, 0)
+    const coinbase = this.headerData.coinbase
+      ? new Address(toBuffer(this.headerData.coinbase))
+      : Address.zero()
+    await rewardAccount(this.vm.stateManager, coinbase, reward)
+  }
+
+  /**
+   * Run and add a transaction to the block being built.
+   * Please note that this modifies the state of the VM.
+   * Throws if the transaction's gasLimit is greater than
+   * the remaining gas in the block.
+   */
+  async addTransaction(tx: TypedTransaction) {
+    this.checkStatus()
+
+    if (!this.checkpointed) {
+      await this.vm.stateManager.checkpoint()
+      this.checkpointed = true
+    }
+
+    // According to the Yellow Paper, a transaction's gas limit
+    // cannot be greater than the remaining gas in the block
+    const blockGasLimit = new BN(toBuffer(this.headerData.gasLimit))
+    const blockGasRemaining = blockGasLimit.sub(this.gasUsed())
+    if (tx.gasLimit.gt(blockGasRemaining)) {
+      throw new Error('tx has a higher gas limit than the remaining gas in the block')
+    }
+
+    const result = await this.vm.runTx({ tx })
+    this.transactions.push(tx)
+    this.transactionResults.push(result)
+    return result
+  }
+
+  /**
+   * Reverts the checkpoint on the StateManager to reset the state from any transactions that have been run.
+   */
+  async revert() {
+    this.checkStatus()
+    if (this.checkpointed) {
+      await this.vm.stateManager.revert()
+      this.reverted = true
+    }
+  }
+
+  /**
+   * This method returns the finalized block.
+   * It also:
+   *  - Assigns the reward for miner
+   *  - Commits the checkpoint on the StateManager
+   *  - Sets the tip of the VM's blockchain to this block
+   * Optionally seals the block with params:
+   *  - PoW: nonce, mixHash, and difficulty validated with the block number by ethash
+   *  - PoA: seals the block with the private key of the clique signer if provided
+   */
+  async build(sealOpts?: SealBlockOpts) {
+    this.checkStatus()
+    const blockOpts = this.blockOpts
+    const consensusType = this.vm._common.consensusType()
+
+    await this.rewardMiner()
+    await this.vm.stateManager.commit()
+
+    const stateRoot = await this.vm.stateManager.getStateRoot(false)
+    const transactionsTrie = await this.transactionsTrie()
+    const receiptTrie = await this.receiptTrie()
+    const bloom = this.bloom()
+    const gasUsed = this.gasUsed()
+    const timestamp = this.headerData.timestamp ?? Date.now()
```

Comment:
Date.now() is not an int

Reply:
thanks, actually it is a number but in ms so I will divide by 1000 for seconds. fixed in https://github.com/ethereumjs/ethereumjs-monorepo/pull/1158/commits/a2226ccb0722bf331b531a75d204078aabebc86b